File: exercism/haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

earthSeconds :: Float
earthSeconds = 31557600

earthYear :: Planet -> Float
earthYear Mercury = 0.2408467
earthYear Venus = 0.61519726
earthYear Earth = 1.0
earthYear Mars = 1.8808158
earthYear Jupiter = 11.862615
earthYear Saturn = 29.447498
earthYear Uranus = 84.016846
earthYear Neptune = 164.79132

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / (earthYear planet * earthSeconds)

File: exercism/elixir/space-age/space_age.exs
defmodule SpaceAge do
  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
                | :saturn | :uranus | :neptune

  @spec to_earth_years(pos_integer) :: float
  def to_earth_years(seconds) do
    seconds / 31557600
  end

  @spec age_on(planet, pos_integer) :: float
  def age_on(planet, seconds) do
    case planet do
      :mercury -> to_earth_years(seconds) / 0.2408467
      :venus -> to_earth_years(seconds) / 0.6151972
      :earth -> to_earth_years(seconds)
      :mars -> to_earth_years(seconds) / 1.8808158
      :jupiter -> to_earth_years(seconds) / 11.862615
      :saturn -> to_earth_years(seconds) / 29.447498
      :uranus -> to_earth_years(seconds) / 84.016846
      :neptune -> to_earth_years(seconds) / 164.79132
    end
  end
end
